In [2]:
import pandas as pd
import numpy as np
from IPython.display import display  

## Cargamos CSV y realizamos visualización

In [3]:
df = pd.read_csv("data.csv", index_col=0) 

In [4]:
display(df.sample(15))

age attrition businesstravel    dailyrate department  distancefromhome  \
438   45        No     non-travel   452.420635        NaN                26   
1568  45        No  travel_rarely   495.753968        NaN                 2   
612   48        No            NaN  1742.103175        NaN                 1   
1656  41        No            NaN  1712.182540        NaN                 2   
461   46       Yes  travel_rarely   556.256661        NaN               -19   
1638  48        No  travel_rarely   417.960317        NaN                 2   
598   32        No  travel_rarely   652.269841        NaN                 9   
1396  39       Yes            NaN   290.035510        NaN                 2   
122   30        No  travel_rarely   556.256661     Sales                  5   
963   38        No  travel_rarely   299.103175     Sales                  9   
856   31        No            NaN  1032.487286        NaN                 3   
1161  29       Yes  travel_rarely   493.896825        NaN               -38   
1513  31        No  travel_rarely   661.349206        NaN                 2   
1391  18        No     non-travel   166.214286        NaN                 1   
1183  31       Yes            NaN   436.738095     Sales                 13   

      education   educationfield  employeecount  employeenumber  ...  \
438           3        Marketing              1             439  ...   
1568          3            Other              1            1569  ...   
612           4              NaN              1             613  ...   
1656          5              NaN              1               9  ...   
461           2          Medical              1             462  ...   
1638          5              NaN              1            1568  ...   
598           4              NaN              1             599  ...   
1396          3              NaN              1            1397  ...   
122           3        Marketing              1             123  ...   
963           3        Marketing              1             964  ...   
856           4          Medical              1             857  ...   
1161          4              NaN              1            1162  ...   
1513          3  Human Resources              1            1514  ...   
1391          3              NaN              1            1392  ...   
1183          4          Medical              1            1184  ...   

      yearsatcompany  yearsincurrentrole  yearssincelastpromotion  \
438               10                 NaN                        4   
1568               8                 NaN                        3   
612                5                 NaN                        2   
1656              18                 NaN                       11   
461                9                 7,0                        3   
1638               9                 NaN                        6   
598                6                 NaN                        0   
1396               1                 NaN                        0   
122               10                 NaN                        1   
963                2                 NaN                        1   
856                2                 NaN                        2   
1161               2                 NaN                        2   
1513               2                 NaN                        1   
1391               0                 NaN                        0   
1183               3                 NaN                        1   

      yearswithcurrmanager  sameasmonthlyincome datebirth      salary  \
438                      5             3654,17$      1978         NaN   
1568                     7                  NaN      1978   48050,00$   
612                      1                  NaN      1975  168850,00$   
1656                     8            13829,17$      1982  165950,00$   
461                      7             4492,84$      1977         NaN   
1638                     7             3375,

In [5]:
df.shape
print(f'Tenemos {df.shape[0]} filas y {df.shape[1]} columnas')

Tenemos 1678 filas y 41 columnas


In [6]:
df.columns

Index(['age', 'attrition', 'businesstravel', 'dailyrate', 'department',
       'distancefromhome', 'education', 'educationfield', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'gender', 'hourlyrate',
       'jobinvolvement', 'joblevel', 'jobrole', 'jobsatisfaction',
       'maritalstatus', 'monthlyincome', 'monthlyrate', 'numcompaniesworked',
       'over18', 'overtime', 'percentsalaryhike', 'performancerating',
       'relationshipsatisfaction', 'standardhours', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   object 
 1   attrition                 1678 non-null   object 
 2   businesstravel            877 non-null    object 
 3   dailyrate                 1678 non-null   float64
 4   department                312 non-null    object 
 5   distancefromhome          1678 non-null   int64  
 6   education                 1678 non-null   int64  
 7   educationfield            904 non-null    object 
 8   employeecount             1678 non-null   int64  
 9   employeenumber            1678 non-null   int64  
 10  environmentsatisfaction   1678 non-null   int64  
 11  gender                    1678 non-null   int64  
 12  hourlyrate                411 non-null    float64
 13  jobinvolvement            1678 non-null   int64  
 14  joblevel     

## Porcentaje de valores nulos en cada columna

In [8]:
(df.isnull().sum() / len(df)) * 100

age                           0.000000
attrition                     0.000000
businesstravel               47.735399
dailyrate                     0.000000
department                   81.406436
distancefromhome              0.000000
education                     0.000000
educationfield               46.126341
employeecount                 0.000000
employeenumber                0.000000
environmentsatisfaction       0.000000
gender                        0.000000
hourlyrate                   75.506555
jobinvolvement                0.000000
joblevel                      0.000000
jobrole                       0.000000
jobsatisfaction               0.000000
maritalstatus                40.226460
monthlyincome                29.141836
monthlyrate                   0.000000
numcompaniesworked            0.000000
over18                       55.899881
overtime                     41.477950
percentsalaryhike             0.000000
performancerating            11.918951
relationshipsatisfaction 

## Dividimos la limpieza y transformación de columnas según el tipo de dato (categóricas/numéricas).

### TRANSFORMACIÓN COLUMNAS CATEGÓRICAS

In [9]:
columns_object = df.select_dtypes(include = object)

In [10]:
columns_object.shape

(1678, 20)

In [11]:
columns_object.columns

Index(['age', 'attrition', 'businesstravel', 'department', 'educationfield',
       'jobrole', 'maritalstatus', 'monthlyincome', 'monthlyrate', 'over18',
       'overtime', 'performancerating', 'standardhours', 'totalworkingyears',
       'worklifebalance', 'yearsincurrentrole', 'sameasmonthlyincome',
       'salary', 'roledepartament', 'remotework'],
      dtype='object')

In [12]:
columnas_categ = ['age', 'attrition', 'businesstravel', 'department', 'educationfield',
       'jobrole', 'maritalstatus', 'monthlyincome', 'monthlyrate', 'over18',
       'overtime', 'performancerating', 'standardhours', 'totalworkingyears',
       'worklifebalance', 'yearsincurrentrole', 'sameasmonthlyincome',
       'salary', 'roledepartament', 'remotework']

### Visualizamos los valores en cada columna para comprobar en mayor medida qué transformaciones requieren los datos

In [13]:
for col in columnas_categ:
    print(f"Valores únicos en '{col}':")
    print(df[col].unique(), "\n")

Valores únicos en 'age':
['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty-seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty-eight' '29' '31' '54' '57' '27' 'thirty-six' '28' '26'
 'fifty-five' '25' 'fifty-two' 'thirty-one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty-six' '19' 'thirty-seven' '18' 'thirty-two' 'twenty-four'] 

Valores únicos en 'attrition':
['No' 'Yes'] 

Valores únicos en 'businesstravel':
[nan 'travel_rarely' 'travel_frequently' 'non-travel'] 

Valores únicos en 'department':
[nan ' Research & Development ' ' Sales ' ' Human Resources '] 

Valores únicos en 'educationfield':
[nan 'Life Sciences' 'Technical Degree' 'Medical' 'Other' 'Marketing'
 'Human Resources'] 

Valores únicos en 'jobrole':
[' resEArch DIREcToR ' ' ManAGeR ' ' ManaGER ' ... ' sAlES ExECUTivE '
 ' SaLes ExecUtIVe ' ' mAnUfactURInG DiRECTOr '] 

Valores únicos en 'maritalstatus':
[nan 'Married' 'Divorced' 'Single' 'divorced' 'Marreid'] 



In [14]:
df[columnas_categ].info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   age                  1678 non-null   object
 1   attrition            1678 non-null   object
 2   businesstravel       877 non-null    object
 3   department           312 non-null    object
 4   educationfield       904 non-null    object
 5   jobrole              1678 non-null   object
 6   maritalstatus        1003 non-null   object
 7   monthlyincome        1189 non-null   object
 8   monthlyrate          1678 non-null   object
 9   over18               740 non-null    object
 10  overtime             982 non-null    object
 11  performancerating    1478 non-null   object
 12  standardhours        1327 non-null   object
 13  totalworkingyears    1129 non-null   object
 14  worklifebalance      1564 non-null   object
 15  yearsincurrentrole   35 non-null     object
 16  sameasmonth

#### Función de transformación para columnas categóricas

In [15]:
def limpieza(columna):
    
    try: 
        columna = columna.str.strip()
        columna = columna.str.lower()
        columna= columna.str.replace(',','.', regex = False)
        columna = columna.str.replace('$','', regex = False)
        columna = columna.str.replace('-','_', regex = False)
        columna = columna.str.replace(' ','_', regex = False)
    
        return columna
    
    except:
        print(f'Error en la columna{columna}')
        return columna

In [16]:
df[columnas_categ] = df[columnas_categ].apply(limpieza)

In [17]:
for col in columnas_categ:
    print(f"Valores únicos en '{col}':")
    print(df[col].unique(), "\n")

Valores únicos en 'age':
['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty_seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty_eight' '29' '31' '54' '57' '27' 'thirty_six' '28' '26'
 'fifty_five' '25' 'fifty_two' 'thirty_one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty_six' '19' 'thirty_seven' '18' 'thirty_two' 'twenty_four'] 

Valores únicos en 'attrition':
['no' 'yes'] 

Valores únicos en 'businesstravel':
[nan 'travel_rarely' 'travel_frequently' 'non_travel'] 

Valores únicos en 'department':
[nan 'research_&_development' 'sales' 'human_resources'] 

Valores únicos en 'educationfield':
[nan 'life_sciences' 'technical_degree' 'medical' 'other' 'marketing'
 'human_resources'] 

Valores únicos en 'jobrole':
['research_director' 'manager' 'sales_executive' 'manufacturing_director'
 'research_scientist' 'healthcare_representative' 'laboratory_technician'
 'sales_representative' 'human_resources'] 

Valores únicos en 'maritalsta

In [18]:
df['maritalstatus']=df['maritalstatus'].str.replace ('marreid', 'married')
df['maritalstatus'].unique()

array([nan, 'married', 'divorced', 'single'], dtype=object)

In [19]:
df['remotework']=df['remotework'].str.replace ('1', 'yes')
df['remotework']=df['remotework'].str.replace ('true', 'yes')
df['remotework']=df['remotework'].str.replace ('0', 'no')
df['remotework']=df['remotework'].str.replace ('false', 'no')
df['remotework'].unique()

array(['yes', 'no'], dtype=object)

In [20]:
columnas_categ_num_int = ['age', 'performancerating', 'totalworkingyears','worklifebalance', 'yearsincurrentrole', 'salary']
columnas_categ_num_float = ['monthlyincome', 'monthlyrate', 'sameasmonthlyincome']

### Comprobamos si hay valores no numéricos antes de convertir el tipo de dato

In [21]:
for columna in columnas_categ_num_int:
    no_numericos = df[~df[columna].astype(str).str.isnumeric()][columna].unique()
    print(f"Valores no numéricos en '{columna}': {no_numericos}")

Valores no numéricos en 'age': ['forty_seven' 'fifty_eight' 'thirty_six' 'fifty_five' 'fifty_two'
 'thirty_one' 'thirty' 'twenty_six' 'thirty_seven' 'thirty_two'
 'twenty_four']
Valores no numéricos en 'performancerating': ['3.0' '4.0' nan]
Valores no numéricos en 'totalworkingyears': [nan '34.0' '22.0' '28.0' '20.0' '21.0' '33.0' '40.0' '18.0' '25.0' '15.0'
 '17.0' '26.0' '16.0' '24.0' '14.0' '23.0' '27.0' '19.0' '11.0' '38.0'
 '37.0' '13.0' '12.0' '29.0' '10.0' '36.0' '35.0' '9.0' '31.0' '32.0'
 '8.0' '7.0' '30.0' '6.0' '5.0' '4.0' '3.0' '2.0' '1.0' '0.0']
Valores no numéricos en 'worklifebalance': ['3.0' nan '2.0' '4.0' '1.0']
Valores no numéricos en 'yearsincurrentrole': [nan '13.0' '12.0' '11.0' '7.0' '6.0' '4.0' '3.0' '2.0' '1.0' '0.0']
Valores no numéricos en 'salary': ['195370.00' '199990.00' '192320.00' '171690.00' nan '100071.84'
 '53914.11' '165950.00' '199730.00' '134020.00' '132060.00' '195450.00'
 '180410.00' '192460.00' '107480.00' '167520.00' '62010.00' '198450.00'
 '40

### Conversión de los números enteros

#### Cambiamos las edades escritas en palabra a número

In [22]:
word_to_num = {
    "twenty": 20, "twenty_one": 21, "twenty_two": 22, "twenty_three": 23, "twenty_four": 24,
    "twenty_five": 25, "twenty_six": 26, "twenty_seven": 27, "twenty_eight": 28, "twenty_nine": 29,
    "thirty": 30, "thirty_one": 31, "thirty_two": 32, "thirty_three": 33, "thirty_four": 34,
    "thirty_five": 35, "thirty_six": 36, "thirty_seven": 37, "thirty_eight": 38, "thirty_nine": 39,
    "forty": 40, "forty_one": 41, "forty_two": 42, "forty_three": 43, "forty_four": 44,
    "forty_five": 45, "forty_six": 46, "forty_seven": 47, "forty_eight": 48, "forty_nine": 49,
    "fifty": 50, "fifty_one": 51, "fifty_two": 52, "fifty_three": 53, "fifty_four": 54,
    "fifty_five": 55, "fifty_six": 56, "fifty_seven": 57, "fifty_eight": 58, "fifty_nine": 59
}

df["age"] = df["age"].replace(word_to_num)

#### Quitamos las comillas de los números para que no sean reconocidos como strings

In [23]:
df[columnas_categ_num_int]= df[columnas_categ_num_int].replace('"','')
df[columnas_categ_num_int]

age performancerating totalworkingyears worklifebalance  \
0     51               3.0               NaN             3.0   
1     52               3.0              34.0             3.0   
2     42               3.0              22.0             NaN   
3     47               3.0               NaN             NaN   
4     46               3.0               NaN             3.0   
...   ..               ...               ...             ...   
1673  43               3.0               NaN             3.0   
1674  47               3.0              27.0             3.0   
1675  29               3.0               6.0             3.0   
1676  47               NaN               NaN             3.0   
1677  32               3.0              10.0             3.0   

     yearsincurrentrole     salary  
0                   NaN  195370.00  
1                   NaN  199990.00  
2                   NaN  192320.00  
3                   NaN  171690.00  
4                   NaN        NaN  
...                 ...        ...  
1673                NaN        NaN  
1674                NaN  191324.62  
1675                NaN   28111.13  
1676                NaN  100071.84  
1677                NaN   53914.11  

[1678 rows x 6 columns]

#### Redondeamos para que no aparezban los num.00 y evitar sea considerado un float

In [24]:
df[columnas_categ_num_int] = df[columnas_categ_num_int].apply(pd.to_numeric, errors='coerce')

In [25]:
df[columnas_categ_num_int] = df[columnas_categ_num_int].round().astype('Int64')

In [26]:
df[columnas_categ_num_int]

age  performancerating  totalworkingyears  worklifebalance  \
0      51                  3               <NA>                3   
1      52                  3                 34                3   
2      42                  3                 22             <NA>   
3      47                  3               <NA>             <NA>   
4      46                  3               <NA>                3   
...   ...                ...                ...              ...   
1673   43                  3               <NA>                3   
1674   47                  3                 27                3   
1675   29                  3                  6                3   
1676   47               <NA>               <NA>                3   
1677   32                  3                 10                3   

      yearsincurrentrole  salary  
0                   <NA>  195370  
1                   <NA>  199990  
2                   <NA>  192320  
3                   <NA>  171690  
4                   <NA>    <NA>  
...                  ...     ...  
1673                <NA>    <NA>  
1674                <NA>  191325  
1675                <NA>   28111  
1676                <NA>  100072  
1677                <NA>   53914  

[1678 rows x 6 columns]

#### Conversión de los números decimales

In [27]:
df[columnas_categ_num_float] = (df[columnas_categ_num_float]).astype(float)

In [28]:
df[columnas_categ_num_float] 

monthlyincome  monthlyrate  sameasmonthlyincome
0          16280.83     42330.17             16280.83
1               NaN     43331.17                  NaN
2               NaN     41669.33                  NaN
3          14307.50     37199.50             14307.50
4          12783.92     33238.20             12783.92
...             ...          ...                  ...
1673        3949.17     10267.83              3949.17
1674       15943.72     41453.67             15943.72
1675            NaN      6090.75                  NaN
1676        8339.32     21682.23              8339.32
1677            NaN     11681.39                  NaN

[1678 rows x 3 columns]

📌 ¿Por qué NaN aparece en float, pero < NA > en int?

1️⃣ En columnas float64, Pandas usa NaN, que es un valor especial de numpy (np.nan).

2️⃣ En columnas int64, no pueden existir nulos (los enteros clásicos no admiten NaN).

3️⃣ En columnas Int64 (nullable), Pandas usa < NA >, que es su propia representación de valores 

In [29]:
df[columnas_categ].info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   age                  1678 non-null   Int64  
 1   attrition            1678 non-null   object 
 2   businesstravel       877 non-null    object 
 3   department           312 non-null    object 
 4   educationfield       904 non-null    object 
 5   jobrole              1678 non-null   object 
 6   maritalstatus        1003 non-null   object 
 7   monthlyincome        1189 non-null   float64
 8   monthlyrate          1678 non-null   float64
 9   over18               740 non-null    object 
 10  overtime             982 non-null    object 
 11  performancerating    1478 non-null   Int64  
 12  standardhours        1327 non-null   object 
 13  totalworkingyears    1129 non-null   Int64  
 14  worklifebalance      1564 non-null   Int64  
 15  yearsincurrentrole   35 non-null     Int64 

### TRANSFORMACIÓN COLUMNAS NUMÉRICAS

In [30]:
# Solo columnas numéricas (int y float)
df_num = df.select_dtypes(exclude="object")
display(df_num.head())

age    dailyrate  distancefromhome  education  employeecount  \
0   51  2015.722222                 6          3              1   
1   52  2063.388889                 1          4              1   
2   42  1984.253968                 4          2              1   
3   47  1771.404762                 2          4              1   
4   46  1582.771346                 3          3              1   

   employeenumber  environmentsatisfaction  gender  hourlyrate  \
0               1                        1       0         NaN   
1               2                        3       0         NaN   
2               3                        3       0         NaN   
3               4                        1       1         NaN   
4               5                        1       1         NaN   

   jobinvolvement  ...  trainingtimeslastyear  worklifebalance  \
0               3  ...                      5                3   
1               2  ...                      5                3   
2               3  ...                      3             <NA>   
3               3  ...                      2             <NA>   
4               4  ...                      5                3   

   yearsatcompany  yearsincurrentrole  yearssincelastpromotion  \
0              20                <NA>                       15   
1              33                <NA>                       11   
2              22                <NA>                       11   
3              20                <NA>                        5   
4              19                <NA>                        2   

   yearswithcurrmanager  sameasmonthlyincome  datebirth  salary  \
0                    15             16280.83       1972  195370   
1                     9                  NaN       1971  199990   
2                    15                  NaN       1981  192320   
3                     6             14307.50       1976  171690   
4                     8             12783.92       1977    <NA>   

   numberchildren  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 30 columns]

In [31]:
df_num.shape

(1678, 30)

In [32]:
(df_num.isnull().sum() / len(df)) * 100

age                           0.000000
dailyrate                     0.000000
distancefromhome              0.000000
education                     0.000000
employeecount                 0.000000
employeenumber                0.000000
environmentsatisfaction       0.000000
gender                        0.000000
hourlyrate                   75.506555
jobinvolvement                0.000000
joblevel                      0.000000
jobsatisfaction               0.000000
monthlyincome                29.141836
monthlyrate                   0.000000
numcompaniesworked            0.000000
percentsalaryhike             0.000000
performancerating            11.918951
relationshipsatisfaction      0.000000
stockoptionlevel              0.000000
totalworkingyears            32.717521
trainingtimeslastyear         0.000000
worklifebalance               6.793802
yearsatcompany                0.000000
yearsincurrentrole           97.914184
yearssincelastpromotion       0.000000
yearswithcurrmanager     

In [33]:
df_num.dtypes

age                           Int64
dailyrate                   float64
distancefromhome              int64
education                     int64
employeecount                 int64
employeenumber                int64
environmentsatisfaction       int64
gender                        int64
hourlyrate                  float64
jobinvolvement                int64
joblevel                      int64
jobsatisfaction               int64
monthlyincome               float64
monthlyrate                 float64
numcompaniesworked            int64
percentsalaryhike             int64
performancerating             Int64
relationshipsatisfaction      int64
stockoptionlevel              int64
totalworkingyears             Int64
trainingtimeslastyear         int64
worklifebalance               Int64
yearsatcompany                int64
yearsincurrentrole            Int64
yearssincelastpromotion       int64
yearswithcurrmanager          int64
sameasmonthlyincome         float64
datebirth                   

In [34]:
df_num.describe().T

count          mean           std         min  \
age                       1678.0      36.94994      9.155706        18.0   
dailyrate                 1678.0    668.079714    470.787298  104.103175   
distancefromhome          1678.0      4.504172     14.652066       -49.0   
education                 1678.0      2.932658       1.02427         1.0   
employeecount             1678.0           1.0           0.0         1.0   
employeenumber            1678.0    809.859952    467.084867         1.0   
environmentsatisfaction   1678.0      4.264005      6.912695         1.0   
gender                    1678.0      0.398689      0.489774         0.0   
hourlyrate                 411.0     83.140768     57.272101   13.012897   
jobinvolvement            1678.0      2.740763      0.710359         1.0   
joblevel                  1678.0      2.064362      1.099425         1.0   
jobsatisfaction           1678.0      2.743147      1.105172         1.0   
monthlyincome             1189.0   5676.787065   6157.859086      1000.0   
monthlyrate               1678.0  14029.674648   9886.532728     2186.17   
numcompaniesworked        1678.0      2.670441      2.501133         0.0   
percentsalaryhike         1678.0      15.15435      3.649607        11.0   
performancerating         1478.0      3.149526      0.356727         3.0   
relationshipsatisfaction  1678.0      2.700238      1.079494         1.0   
stockoptionlevel          1678.0      0.787843      0.843252         0.0   
totalworkingyears         1129.0      11.35961      7.807298         0.0   
trainingtimeslastyear     1678.0       2.79559      1.288657         0.0   
worklifebalance           1564.0      2.759591      0.703527         1.0   
yearsatcompany            1678.0      7.124553      6.173837         0.0   
yearsincurrentrole          35.0      3.971429      3.584983         0.0   
yearssincelastpromotion   1678.0      2.241955      3.246554         0.0   
yearswithcurrmanager      1678.0      4.200238      3.554224         0.0   
sameasmonthlyincome       1189.0   5676.787065   6157.859086      1000.0   
datebirth                 1678.0    1986.04708      9.154908      1963.0   
salary                    1393.0  65070.885858  45849.562585     10090.0   
numberchildren               0.0           NaN           NaN         NaN   

                                25%         50%         75%          max  
age                            30.0        36.0        43.0         60.0  
dailyrate                 290.03551  556.256661  971.956349  2063.388889  
distancefromhome                2.0         5.0        11.0         29.0  
education                       2.0         3.0         4.0          5.0  
employeecount                   1.0         1.0         1.0          1.0  
employeenumber               403.25       813.5     1215.75       1614.0  
environmentsatisfaction         2.0         3.0         4.0         49.0  
gender                          0.0         0.0         1.0          1.0  
hourlyrate                36.254439   69.532083  116.987103   255.963294  
jobinvolvement                  2.0         3.0         3.0          4.0  
joblevel                        1.0         2.0         3.0          5.0  
jobsatisfaction                 2.0         3.0         4.0          4.0  
monthlyincome               2342.59     4492.84     7861.67      94083.0  
monthlyrate                 6090.75    11681.39    20411.08     43331.17  
numcompaniesworked              1.0         2.0         4.0          9.0  
percentsalaryhike              12.0        14.0        18.0         25.0  
performancerating               3.0         3.0         3.0          4.0  
relationshipsatisfaction        2.0         3.0         4.0          4.0  
stockoptionlevel                0.0         1.0         1.0          3.0  
totalworkingyears               6.0        10.0        15.0         40.0  
trainingtimeslastyear           2.0         3.0         3.0          6.0  
worklifebalance     

In [35]:
columnas_numericas_trans = ['dailyrate','distancefromhome','education','employeecount','employeenumber','environmentsatisfaction','gender','hourlyrate','jobinvolvement','joblevel', 
                    'jobsatisfaction','numcompaniesworked','percentsalaryhike','relationshipsatisfaction','stockoptionlevel','trainingtimeslastyear','yearsatcompany',
                    'yearssincelastpromotion','yearswithcurrmanager','datebirth']


#### COLUMNA DISTANCEFROMHOME: TIENE VALORES NEGATIVOS

In [36]:
df["distancefromhome"] = df["distancefromhome"].abs()

In [37]:
df["distancefromhome"]

0        6
1        1
2        4
3        2
4        3
        ..
1673    26
1674    26
1675    15
1676     4
1677     2
Name: distancefromhome, Length: 1678, dtype: int64

#### TRANSFORMAMOS LOS DATOS VARIABLES DE GÉNERO SEGÚN INDICACIÓN EMPRESARIAL

In [38]:
df['gender'] = df['gender'].map({0: 'male', 1: 'female'})

In [39]:
df['gender'].unique()

array(['male', 'female'], dtype=object)

In [40]:
# COLUMNA ENVOIREMENTSATISFACTION.
# ESTA COLUMNA VA DE 1 A 5, PERO APARECE UN VALOR MAXIMO DE 49

valores_validos = [1, 2, 3, 4, 5] # creo una lista con valores validos (escala de 1 a 5)

nueva_satisfaccion = [] # creo una lista vacia 
for valor in df['environmentsatisfaction']:
    if valor in valores_validos:  # por cada valor de la columna, verifica si el valor esta en la lista de validos
        nueva_satisfaccion.append(valor)  # si el valor es valido lo apendeo a la lista vacia
    else:
        nueva_satisfaccion.append(None) # sino dame None.

df['environmentsatisfaction'] = nueva_satisfaccion  # reemplaxzo la columna original por la mnueva


In [41]:
df['environmentsatisfaction'].sample(5)

1418    1.0
939     1.0
329     1.0
272     1.0
510     3.0
Name: environmentsatisfaction, dtype: float64

### LIMPIEZA FINALIZADA. 

In [42]:
df.sample(10)

age attrition     businesstravel    dailyrate              department  \
1507   40        no      travel_rarely  1582.771346  research_&_development   
1419   31        no                NaN   556.256661                     NaN   
1108   32        no  travel_frequently   354.198413                     NaN   
1335   32       yes      travel_rarely   290.035510                     NaN   
614    35        no      travel_rarely   446.023810                     NaN   
115    34        no                NaN   556.256661                     NaN   
799    34        no      travel_rarely   290.035510                     NaN   
206    27        no                NaN  1321.460317                     NaN   
531    31        no      travel_rarely   556.256661                   sales   
969    22       yes      travel_rarely   556.256661                     NaN   

      distancefromhome  education    educationfield  employeecount  \
1507                 1          4               NaN              1   
1419                 5          4     life_sciences              1   
1108                10          3               NaN              1   
1335                25          4               NaN              1   
614                  8          4  technical_degree              1   
115                 39          4     life_sciences              1   
799                 38          2               NaN              1   
206                  5          1               NaN              1   
531                  2          1     life_sciences              1   
969                  3          1     life_sciences              1   

      employeenumber  ...  yearsatcompany yearsincurrentrole  \
1507            1508  ...              19               <NA>   
1419            1420  ...              10               <NA>   
1108            1109  ...               5               <NA>   
1335            1336  ...               1               <NA>   
614              615  ...               5               <NA>   
115              116  ...              11               <NA>   
799              800  ...               2               <NA>   
206              207  ...               9               <NA>   
531              532  ...               6               <NA>   
969              970  ...               3               <NA>   

      yearssincelastpromotion  yearswithcurrmanager  sameasmonthlyincome  \
1507                       11                    16             12783.92   
1419                        0                     2              4492.84   
1108                        1                     3              2860.83   
1335                        0                     1              2342.59   
614                         1                     4                  NaN   
115                         5                     9              4492.84   
799                         1                     2              2342.59   
206                         0                     8             10673.33   
531                         0                     5                  NaN   
969                         0                     2              4492.84   

     datebirth  salary                     roledepartament  numberchildren  \
1507      1983  153407  manager_____research_&_development             NaN   
1419      1992   53914                                 NaN             NaN   
1108      1991    <NA>                                 NaN             NaN   
1335      1991   28111                                 NaN             NaN   
614       1988   43230                                 NaN             NaN   
115       1989   53914                                 NaN             NaN   
799       1989    <NA>                                 NaN             NaN   
206       1996  128080                                 NaN             NaN   
531       1992   53914           sales_executive_____sales             NaN   
969       2001   53914                     

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   Int64  
 1   attrition                 1678 non-null   object 
 2   businesstravel            877 non-null    object 
 3   dailyrate                 1678 non-null   float64
 4   department                312 non-null    object 
 5   distancefromhome          1678 non-null   int64  
 6   education                 1678 non-null   int64  
 7   educationfield            904 non-null    object 
 8   employeecount             1678 non-null   int64  
 9   employeenumber            1678 non-null   int64  
 10  environmentsatisfaction   1576 non-null   float64
 11  gender                    1678 non-null   object 
 12  hourlyrate                411 non-null    float64
 13  jobinvolvement            1678 non-null   int64  
 14  joblevel     

### RENOMBRAMOS LAS COLUMNAS PARA HOMOGENIZAR Y HACER MÁS VISIBLES SUS NOMBRES

In [44]:
df.columns.tolist()

['age',
 'attrition',
 'businesstravel',
 'dailyrate',
 'department',
 'distancefromhome',
 'education',
 'educationfield',
 'employeecount',
 'employeenumber',
 'environmentsatisfaction',
 'gender',
 'hourlyrate',
 'jobinvolvement',
 'joblevel',
 'jobrole',
 'jobsatisfaction',
 'maritalstatus',
 'monthlyincome',
 'monthlyrate',
 'numcompaniesworked',
 'over18',
 'overtime',
 'percentsalaryhike',
 'performancerating',
 'relationshipsatisfaction',
 'standardhours',
 'stockoptionlevel',
 'totalworkingyears',
 'trainingtimeslastyear',
 'worklifebalance',
 'yearsatcompany',
 'yearsincurrentrole',
 'yearssincelastpromotion',
 'yearswithcurrmanager',
 'sameasmonthlyincome',
 'datebirth',
 'salary',
 'roledepartament',
 'numberchildren',
 'remotework']

In [45]:
# primero renombro una columna y luego hago una funcion

# Renombro la columna 'jobinvolvement' a 'job_involvement_level'
df.rename(columns={'jobinvolvement': 'job_involvement_level'}, inplace=True)

# Verifico que se ha renombrado correctamente
print(df.columns) 


Index(['age', 'attrition', 'businesstravel', 'dailyrate', 'department',
       'distancefromhome', 'education', 'educationfield', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'gender', 'hourlyrate',
       'job_involvement_level', 'joblevel', 'jobrole', 'jobsatisfaction',
       'maritalstatus', 'monthlyincome', 'monthlyrate', 'numcompaniesworked',
       'over18', 'overtime', 'percentsalaryhike', 'performancerating',
       'relationshipsatisfaction', 'standardhours', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')


In [46]:
df.shape

(1678, 41)

In [47]:
df.columns

Index(['age', 'attrition', 'businesstravel', 'dailyrate', 'department',
       'distancefromhome', 'education', 'educationfield', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'gender', 'hourlyrate',
       'job_involvement_level', 'joblevel', 'jobrole', 'jobsatisfaction',
       'maritalstatus', 'monthlyincome', 'monthlyrate', 'numcompaniesworked',
       'over18', 'overtime', 'percentsalaryhike', 'performancerating',
       'relationshipsatisfaction', 'standardhours', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'roledepartament', 'numberchildren', 'remotework'],
      dtype='object')

In [48]:
mapa_columnas = {
    'businesstravel': 'business_travel', 
    'dailyrate': 'daily_rate',
    'distancefromhome': 'distance_from_home',
    'educationfield': 'education_field', 
    'employeecount': 'employee_count',
    'employeenumber': 'employee_number',
    'environmentsatisfaction': 'environment_satisfaction', 
    'hourlyrate': 'hourly_rate',
    'joblevel': 'job_level', 
    'jobrole': 'job_role', 
    'jobsatisfaction': 'job_satisfaction',
    'maritalstatus': 'marital_status', 
    'monthlyincome': 'monthly_income',
    'monthlyrate': 'monthly_rate', 
    'numcompaniesworked': 'num_companies_worked',
    'over18': 'over_18', 
    'overtime': 'overtime', 
    'percentsalaryhike': 'percent_salary_hike', 
    'performancerating': 'performance_rating',
    'relationshipsatisfaction': 'relationship_satisfaction', 
    'standardhours': 'standard_hours', 
    'stockoptionlevel': 'stock_option_level',
    'totalworkingyears': 'total_working_years', 
    'trainingtimeslastyear': 'training_times_last_year', 
    'worklifebalance': 'work_life_balance',
    'yearsatcompany': 'years_at_company', 
    'yearsincurrentrole': 'years_in_current_role', 
    'yearssincelastpromotion': 'years_since_last_promotion',
    'yearswithcurrmanager': 'years_with_curr_manager', 
    'sameasmonthlyincome': 'same_as_monthly_income', 
    'datebirth': 'date_birth', 
    'roledepartament': 'role_department', 
    'numberchildren': 'number_children', 
    'remotework': 'remote_work'
}

In [49]:
df.rename(columns=mapa_columnas, inplace=True)

In [50]:
df.columns

Index(['age', 'attrition', 'business_travel', 'daily_rate', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_count',
       'employee_number', 'environment_satisfaction', 'gender', 'hourly_rate',
       'job_involvement_level', 'job_level', 'job_role', 'job_satisfaction',
       'marital_status', 'monthly_income', 'monthly_rate',
       'num_companies_worked', 'over_18', 'overtime', 'percent_salary_hike',
       'performance_rating', 'relationship_satisfaction', 'standard_hours',
       'stock_option_level', 'total_working_years', 'training_times_last_year',
       'work_life_balance', 'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'years_with_curr_manager',
       'same_as_monthly_income', 'date_birth', 'salary', 'role_department',
       'number_children', 'remote_work'],
      dtype='object')

### Revisiones finales

In [51]:
df['environment_satisfaction'].dtypes

dtype('float64')

In [52]:
df['environment_satisfaction'].astype('Int64')

0       1
1       3
2       3
3       1
4       1
       ..
1673    2
1674    4
1675    3
1676    3
1677    4
Name: environment_satisfaction, Length: 1678, dtype: Int64

Revisamos valores duplicados

In [53]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1673     True
1674     True
1675     True
1676     True
1677     True
Length: 1678, dtype: bool

In [54]:
df.loc[df.duplicated(keep=False)]

age attrition business_travel   daily_rate department  \
8      41        no             NaN  1712.182540        NaN   
60     36        no             NaN   610.174603        NaN   
75     47        no   travel_rarely  1032.487286        NaN   
107    29        no   travel_rarely  1032.487286        NaN   
111    30        no   travel_rarely  1032.487286        NaN   
...   ...       ...             ...          ...        ...   
1673   43        no             NaN   488.944444        NaN   
1674   47        no             NaN  1973.984127        NaN   
1675   29        no   travel_rarely   290.035510        NaN   
1676   47        no   travel_rarely  1032.487286        NaN   
1677   32        no             NaN   556.256661        NaN   

      distance_from_home  education education_field  employee_count  \
8                      2          5             NaN               1   
60                     5          2             NaN               1   
75                     4          3   life_sciences               1   
107                   21          4   life_sciences               1   
111                    5          3             NaN               1   
...                  ...        ...             ...             ...   
1673                  26          3         medical               1   
1674                  26          4             NaN               1   
1675                  15          3             NaN               1   
1676                   4          3   life_sciences               1   
1677                   2          2   life_sciences               1   

      employee_number  ...  years_at_company years_in_current_role  \
8                   9  ...                18                  <NA>   
60                 61  ...                13                  <NA>   
75                 76  ...                22                  <NA>   
107               108  ...                10                  <NA>   
111               112  ...                10                  <NA>   
...               ...  ...               ...                   ...   
1673              824  ...                 3                  <NA>   
1674             1087  ...                 5                  <NA>   
1675              528  ...                 6                  <NA>   
1676               76  ...                22                  <NA>   
1677              401  ...                 8                  <NA>   

      years_since_last_promotion  years_with_curr_manager  \
8                             11                        8   
60                             3                        7   
75                            14                       10   
107                            8                        8   
111                            7                        4   
...                          ...                      ...   
1673                           1                        2   
1674                           1                        0   
1675                           1                        5   
1676                          14                       10   
1677                           0                        7   

      same_as_monthly_income date_birth  salary role_department  \
8                   13829.17       1982  165950             NaN   
60                   4928.33       1987   59140             NaN   
75                   8339.32       1976  100072             NaN   
107                  8339.32       1994    <NA>             NaN   
111                  8339.32       1993  100072             NaN   
...                      ...        ...     ...             ...   
1673                 3949.17       1980    <NA>             NaN   
1674                15943.72       1976  191325             NaN   
1675                     NaN       1994   28111             NaN   
1676                 8339.32       1976  100072             NaN   
1677                     NaN       1991   53914             NaN   

      number_chil

#### Revisamos columna a columna

In [55]:
columnas_duplicadas = [
    'age', 'attrition', 'business_travel', 'daily_rate', 'department',
    'distance_from_home', 'education', 'education_field', 'employee_count',
    'employee_number', 'years_at_company', 'years_in_current_role',
    'years_since_last_promotion', 'years_with_curr_manager',
    'same_as_monthly_income', 'date_birth', 'salary', 'role_department',
    'number_children', 'remote_work'
]

for col in columnas_duplicadas:
    duplicados = df[col][df.duplicated(subset=[col], keep=False)]
    if not duplicados.empty:
        print(f"Duplicados en la columna '{col}':")
        print(duplicados.unique())  # Muestra solo los valores duplicados únicos
        print("-" * 50)

Duplicados en la columna 'age':
<IntegerArray>
[51, 52, 42, 47, 46, 48, 59, 41, 56, 38, 55, 40, 58, 35, 45, 33, 36, 34, 53,
 43, 60, 32, 37, 49, 39, 50, 44, 30, 29, 31, 54, 57, 27, 28, 26, 25, 24, 23,
 22, 21, 20, 19, 18]
Length: 43, dtype: Int64
--------------------------------------------------
Duplicados en la columna 'attrition':
['no' 'yes']
--------------------------------------------------
Duplicados en la columna 'business_travel':
[nan 'travel_rarely' 'travel_frequently' 'non_travel']
--------------------------------------------------
Duplicados en la columna 'daily_rate':
[1582.77134647 1032.48728618  556.256661   1712.18253968 1973.98412698
  635.76190476  290.03550974  610.17460317 1964.96031746 1652.34126984
  991.81746032 1187.53968254  444.26984127 1758.71428571  238.12698413
  654.84920635 1842.8015873   433.02380952  300.13492063  456.44444444
  578.29365079  601.3015873   253.29365079  521.54761905  483.06349206
  542.38888889  218.62698413  616.36507937  313.03174603

Según el output los valores duplicados porblemáticos que debemos limpiar son los siguientes:

Duplicados en la columna 'employee_number':

[   9   61   76  108  112  159  168  178  191  203  271  293  300  304
  352  369  401  526  528  532  564  633  644  663  672  755  824  828
  853  864  878  932  958  969  977 1030 1048 1056 1060 1087 1158 1179
 1212 1235 1243 1245 1268 1324 1326 1339 1342 1349 1370 1393 1402 1474
 1477 1499 1525 1532 1567 1568 1569 1608]

En este código:

    subset='employee_number' especifica que solo se tenga en cuenta la columna employee_number para detectar duplicados.

    keep=False elimina todas las filas que contienen un número de empleado repetido. Si solo quieres mantener la primera aparición de cada número de empleado, usa keep='first'.

In [56]:
df_sin_duplicados = df.drop_duplicates(subset='employee_number', keep='first')

In [57]:
df_sin_duplicados['employee_number'].duplicated(keep =False)

0       False
1       False
2       False
3       False
4       False
        ...  
1609    False
1610    False
1611    False
1612    False
1613    False
Name: employee_number, Length: 1614, dtype: bool

### Como está todo correcto, ahora sí sobreescribimos nuestro df

In [58]:
df = df.drop_duplicates(subset='employee_number', keep='first')

In [59]:
# Contar los valores únicos
num_valores_unicos = len(df['employee_number'].unique())

# Mostrar la cantidad de valores únicos
print(num_valores_unicos)

1614


In [60]:
df['employee_number'].unique()

array([   1,    2,    3, ..., 1612, 1613, 1614])

In [61]:
# Ajustar para mostrar todos los valores en el array
np.set_printoptions(threshold=np.inf)

# Ver todos los valores únicos de la columna 'employee_number'
print(df['employee_number'].unique())

[   1    2    3    4    5    6    7    8    9   10   11   12   13   14
   15   16   17   18   19   20   21   22   23   24   25   26   27   28
   29   30   31   32   33   34   35   36   37   38   39   40   41   42
   43   44   45   46   47   48   49   50   51   52   53   54   55   56
   57   58   59   60   61   62   63   64   65   66   67   68   69   70
   71   72   73   74   75   76   77   78   79   80   81   82   83   84
   85   86   87   88   89   90   91   92   93   94   95   96   97   98
   99  100  101  102  103  104  105  106  107  108  109  110  111  112
  113  114  115  116  117  118  119  120  121  122  123  124  125  126
  127  128  129  130  131  132  133  134  135  136  137  138  139  140
  141  142  143  144  145  146  147  148  149  150  151  152  153  154
  155  156  157  158  159  160  161  162  163  164  165  166  167  168
  169  170  171  172  173  174  175  176  177  178  179  180  181  182
  183  184  185  186  187  188  189  190  191  192  193  194  195  196
  197 

In [62]:
df.shape

(1614, 41)

### GUARDAMOS EL DATAFRAME FINAL 

In [63]:
# Guardamos el DataFrame limpio en un archivo CSV
df.to_csv('bbdd_data.csv', index=False)